pkg/virtualfs/directory_test.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2020 The Kanister Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package virtualfs
+
+import (
+	"context"
+
+	. "gopkg.in/check.v1"
+)
+
+const (
+	defaultPermissions = 0777
+)
+
+type DirectorySuite struct{}
+
+var _ = Suite(&DirectorySuite{})
+
+func (s *DirectorySuite) TestAddDir(c *C) {
+	sourceDir, err := NewDirectory("root")
+	c.Assert(err, IsNil)
+
+	// New directory
+	dir, err := sourceDir.AddDir("d1", defaultPermissions)
+	c.Assert(err, IsNil)
+	c.Assert(dir.Name(), Equals, "d1")
+
+	// Duplicate directory
+	dir, err = sourceDir.AddDir("d1", defaultPermissions)
+	c.Assert(err, NotNil)
+
+	// Invalid name
+	dir, err = sourceDir.AddDir("/d2", defaultPermissions)
+	c.Assert(err, NotNil)
+}
+
+func (s *DirectorySuite) TestAddAllDirs(c *C) {
+	sourceDir, err := NewDirectory("root")
+	c.Assert(err, IsNil)
+
+	c.Log("Add a directory - root/d1")
+	subdir, err := sourceDir.AddAllDirs("d1/d2", defaultPermissions)
+	c.Assert(err, IsNil)
+	c.Assert(subdir.Name(), Equals, "d2")
+	d1, err := sourceDir.Subdir("d1")
+	c.Assert(err, IsNil)
+	c.Assert(d1, NotNil)
+
+	c.Log("Add third level directory - root/d1/d2")
```

Comment:
This log message seems misleading . The case below is checking whether adding already existing dirs succeeds (does not fail)